### PR TITLE
nixos/wakapi: add database options; gate db creation behind database.createLocally

### PR DESCRIPTION
**Pull request by NotAShelf**
Currently the NixOS module for Wakapi will create the database automagically if the user has database dialect configured in the Wakapi configuration file. By all means, this is undocumented behaviour and an anti-feature.

This MR adds a database.createLocally option that allows the end-user to create auto-creation behaviour, and lays out groundwork for automated database setups for different database dialects supported by Wakapi.

## Description of changes

Adds `services.wakapi.database.createLocally` and `servicees.wakapi.database.dialect` to fine-grain 
database creation behaviour. Fixes case where database would be created unconditionally when the 
user has DB dialect set in Wakapi config.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by NotAShelf** [comment]
CC @cafkafk as the one who oversaw the init PR.

**Comment by NotAShelf** [comment]
Manual build is failing, I'll fix it shortly.

File: nixos/modules/services/web-apps/wakapi.nix [hunk 1]
```diff
@@ -134,6 +188,10 @@ in
         assertion = cfg.smtpPassword != null -> cfg.smtpPasswordFile != null;
         message = "Both `services.wakapi.smtpPassword` `services.wakapi.smtpPasswordFile` should not be set at the same time.";
       }
+      {
+        assertion = cfg.db.createLocally -> cfg.db.dialect != null;
+        message = "`services.wakapi.database.createLocally` is true, but a database dialect is not set!";
```
Comment (jacekpoz) [inline nixos/modules/services/web-apps/wakapi.nix:193]: maybe add some note about only postgres being supported here too?

**Review by imnotpoz — CHANGES_REQUESTED** [review]
one suggestion otherwise looking good

File: nixos/modules/services/web-apps/wakapi.nix (hunk 1 above)
Comment (jacekpoz) [reply nixos/modules/services/web-apps/wakapi.nix:193]: that would also do, the current message implies that you can pick any dialect and `createLocally` will do something

Comment (NotAShelf) [reply nixos/modules/services/web-apps/wakapi.nix:193]: Though null already covers "not postgres". I'm not sure, maybe it's safer to just mention we only support postgresql.

Comment (NotAShelf) [reply nixos/modules/services/web-apps/wakapi.nix:193]: I've added a warning to get around this. It's probably not worth an assertion anyway.

**Review by imnotpoz — APPROVED** [review]

**Comment by NotAShelf** [comment]
@isabelroses this still awaits your merge if you don't mind.

**Comment by NotAShelf** [comment]
Still needs a review.

**Comment by DeeUnderscore** [comment]
As an aside, Wakapi *can* be configured to talk to Postgresql over a Unix domain socket with peer auth. There's a `db.DSN` config option that can be given something like `postgres://wakapi@/wakapi?host=/run/postgresql`. This is undocumented/underdocumented, but was put in explicitly to allow peer auth: https://github.com/muety/wakapi/issues/423

The effect is that the host auth line that we're adding to the postgresql config via this module may be redundant.

**Review by isabelroses — APPROVED** [review]

**Pull request merged**